scipy/stats/tests/test_qmc.py

```diff
@@ -0,0 +1,853 @@
+
+        with pytest.raises(ValueError, match=r"Bounds are not consistent"
+                                             r" a < b"):
+            space = [[0, 0], [1, 1], [0.5, 0.5]]
+            bounds = np.array([[-2, 6], [6, 5]])
+            qmc.scale(space, l_bounds=bounds[0], u_bounds=bounds[1])
+
+        with pytest.raises(ValueError, match=r"Bounds do not have the same"
+                                             r" dimensions"):
+            space = [[0, 0], [1, 1], [0.5, 0.5]]
+            l_bounds, u_bounds = [-2, 0, 2], [6, 5]
+            qmc.scale(space, l_bounds=l_bounds, u_bounds=u_bounds)
+
+        with pytest.raises(ValueError, match=r"Sample dimension is different "
+                                             r"than bounds dimension"):
+            space = [[0, 0], [1, 1], [0.5, 0.5]]
+            bounds = np.array([[-2, 0, 2], [6, 5, 5]])
+            qmc.scale(space, l_bounds=bounds[0], u_bounds=bounds[1])
+
+        with pytest.raises(ValueError, match=r"Sample is not in hypercube"):
+            space = [[0, 0], [1, 1.5], [0.5, 0.5]]
+            bounds = np.array([[-2, 0], [6, 5]])
+            qmc.scale(space, l_bounds=bounds[0], u_bounds=bounds[1])
+
+        with pytest.raises(ValueError, match=r"Sample is out of bounds"):
+            out = [[-2, 0], [6, 5], [8, 2.5]]
+            bounds = np.array([[-2, 0], [6, 5]])
+            qmc.scale(out, l_bounds=bounds[0], u_bounds=bounds[1],
+                      reverse=True)
+
+    def test_discrepancy(self):
+        space_1 = np.array([[1, 3], [2, 6], [3, 2], [4, 5], [5, 1], [6, 4]])
+        space_1 = (2.0 * space_1 - 1.0) / (2.0 * 6.0)
+        space_2 = np.array([[1, 5], [2, 4], [3, 3], [4, 2], [5, 1], [6, 6]])
+        space_2 = (2.0 * space_2 - 1.0) / (2.0 * 6.0)
+
+        # From Fang et al. Design and modeling for computer experiments, 2006
+        assert_allclose(qmc.discrepancy(space_1), 0.0081, atol=1e-4)
+        assert_allclose(qmc.discrepancy(space_2), 0.0105, atol=1e-4)
+
+        # From Zhou Y.-D. et al. Mixture discrepancy for quasi-random point
+        # sets. Journal of Complexity, 29 (3-4), pp. 283-301, 2013.
+        sample = np.array([[2, 1, 1, 2, 2, 2],
+                           [1, 2, 2, 2, 2, 2],
+                           [2, 1, 1, 1, 1, 1],
+                           [1, 1, 1, 1, 2, 2],
+                           [1, 2, 2, 2, 1, 1],
+                           [2, 2, 2, 2, 1, 1],
+                           [2, 2, 2, 1, 2, 2]])
+        sample = (2.0 * sample - 1.0) / (2.0 * 2.0)
+
+        assert_allclose(qmc.discrepancy(sample, method='MD'), 2.5000,
+                        atol=1e-4)
+        assert_allclose(qmc.discrepancy(sample, method='WD'), 1.3680,
+                        atol=1e-4)
+        assert_allclose(qmc.discrepancy(sample, method='CD'), 0.3172,
+                        atol=1e-4)
+        assert_allclose(qmc.discrepancy(sample, method='star'), 0.037451,
+                        atol=1e-4)
```

Comment:
Where did you get this? It's not calculated in [2], is it?

Reply:
I don't remember. I will have a look...

Reply:
So, I found something in *Tim P. et al. "Minimizing the L2 and L∞ star discrepancies of a single point
in the unit hypercube". JCAM, 2005 (doi:10.1016/j.cam.2005.11.005)*. There is a graph at the end of the article which I reproduced:

```python
import numpy as np
from scipy.stats.qmc import discrepancy
import matplotlib.pyplot as plt


fig, ax = plt.subplots()

points = np.linspace(0, 1, 50)
for dim in [8, 16, 32, 64]:
    disc = []
    for point in points:
        disc.append(discrepancy(np.array([[point]*dim]), method='L2-star'))

    ax.plot(points, disc, label=dim)
    
ax.legend()
ax.set_yscale('log')
```

![convergence](https://user-images.githubusercontent.com/23188539/104710937-fec7a980-5720-11eb-85d2-2f61fa683e3f.png)

Reply:
I added the numerical values from Table 1 as a test.

Reply:
Cool.